circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3Copier.java

```diff
@@ -225,22 +226,74 @@ private void applyObjectMetadata(CopyObjectRequest copyObjectRequest) {
     }
   }
 
-  private Metrics gatherAllCopyResults() {
-    AtomicLong bytesReplicated = new AtomicLong(0);
-    registerRunningMetrics(bytesReplicated);
-    for (Copy copyJob : copyJobs) {
+  private void processAllCopyJobs() {
+    List<CopyJob> copyJobsToSubmit = copyJobs;
+    int maxCopyAttempts = s3s3CopierOptions.getMaxCopyAttempts();
+    for (int copyAttempt = 1; copyAttempt <= maxCopyAttempts; copyAttempt++) {
+      LOG
+          .info("Submitting {} copy job(s), attempt {}/{}", copyJobsToSubmit.size(), copyAttempt, maxCopyAttempts);
+      copyJobsToSubmit = submitAndGatherCopyJobs(copyJobsToSubmit);
+      if (copyJobsToSubmit.isEmpty()) {
+        LOG
+            .info("Successfully gathered all copy jobs on attempt {}/{}", copyAttempt, maxCopyAttempts);
+        break;
+      }
+      if (copyAttempt == maxCopyAttempts) {
+        throw new CircusTrainException(copyJobsToSubmit.size() + " job(s) failed the maximum number of copy attempts, " + maxCopyAttempts);
+      }
+      LOG
+          .info("Finished gathering jobs on attempt {}/{}. Retrying {} failed job(s).",
+              copyAttempt,
+              maxCopyAttempts,
+              copyJobsToSubmit.size());
+    }
+  }
+
+  private List<CopyJob> submitAndGatherCopyJobs(List<CopyJob> copyJobsToSubmit) {
+    List<CopyJob> submittedCopyJobs = new ArrayList<>();
+    for (CopyJob copyJob : copyJobsToSubmit) {
+      Copy copy = submitCopyJob(copyJob);
+      CopyJob newCopyJob = new CopyJob(copy, copyJob.getCopyObjectRequest(), copyJob.getTransferStateChangeListener());
+      submittedCopyJobs.add(newCopyJob);
+    }
+    return gatherCopyJobs(submittedCopyJobs);
+  }
+
+  private Copy submitCopyJob(CopyJob copyJob) {
+    CopyObjectRequest copyObjectRequest = copyJob.getCopyObjectRequest();
+    LOG
+        .info("Copying object from '{}/{}' to '{}/{}'", copyObjectRequest.getSourceBucketName(), copyObjectRequest.getSourceKey(),
+            copyObjectRequest.getDestinationBucketName(), copyObjectRequest.getDestinationKey());
+    return transferManager.copy(copyObjectRequest, srcClient, copyJob.getTransferStateChangeListener());
+  }
+
+  private List<CopyJob> gatherCopyJobs(List<CopyJob> copyJobs) {
```

Comment:
probably worth adding javadoc here just to say that the returned `List` contains the failed jobs

Reply:
or have a `GatherCopyJobResult` object that has `getFailedCopyJobs` method. (I am also fine with a line of javadoc.

Reply:
Added a Javadoc to the method